### PR TITLE
feat: change password

### PR DESCRIPTION
**Pull request by andreabadesso**
### Acceptance Criteria
- Added a method to change the password
- It should also change everything that is encrypted on storage
- We should also store its hash on storage

### Security Checklist
- [X] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/wallet.js [hunk 1]
```diff
@@ -723,6 +723,84 @@ const wallet = {
...
+          break;
+        }
+      } catch(e) {}
+    }
+
+    if (pinFound !== -1) {
+      return [true, pinFound];
+    }
+
+    return [false];
```
Comment (msbrogli) [inline src/wallet.js:801]: Should we return `[false, '']` for consistency in the return type?

File: src/wallet.js [hunk 2]
```diff
@@ -723,6 +723,84 @@ const wallet = {
...
+    return true;
+  },
+
+  /**
+   * @return {[boolean, string]} Tuple <success:boolean, pin:string> with the result
+   *
+   * @memberof Wallet
+   * @inner
+   */
+  guessPin() {
```
Comment (msbrogli) [inline src/wallet.js:770]: I'm not sure we should place it in hathor-wallet-lib. It seems a temporary fix needed only in hathor-wallet-mobile. I feel we should keep it there and remove it in a later version.

File: src/wallet.js [hunk 3]
```diff
@@ -723,6 +723,84 @@ const wallet = {
...
+    const accessData = this.getWalletAccessData();
+    const data = accessData.words;
+
+    const decrypt = (data, pin) => CryptoJS
+      .AES
+      .decrypt(data, pin)
+      .toString(CryptoJS.enc.Utf8);
+
+    let pinFound = -1;
+    for (let i = 0; i <= 999999; i++) {
```
Comment (msbrogli) [inline src/wallet.js:781]: Should we block the main loop? I think we should do something like this:

```js
guessPartial(begin, end) {
  :
  for (let i = begin; i <= end; i++) {
    :
  }
  :
}

guessScheduler(begin, step, cb) {
  const end = begin + step;
  ret = guessPartial(begin, end);
  if (ret[0] == false) {
    if (end < 999999) {
      setTimeout(guessScheduler, 0, end + 1, step, cb);
    }
  } else {
    cb(pin);
  }
}

guessPin(cb) {
  guessScheduler(0, 1000, cb);
}

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: src/wallet.js (hunk 3 above)
Comment (msbrogli) [reply src/wallet.js:781]: Maybe we should call `cb(...)` multiple times so the caller can update the interface with progress.

Comment (andreabadesso) [reply src/wallet.js]: @msbrogli This was moved to the hathor-wallet-mobile repo, can you check it there?

File: src/wallet.js (hunk 2 above)
Comment (andreabadesso) [reply src/wallet.js]: I agree, moved to the hathor-wallet-mobile repo and removed the latest commit

File: src/wallet.js (hunk 1 above)
Comment (andreabadesso) [reply src/wallet.js]: I agree, done in the hathor-wallet-mobile repo

File: src/wallet.js [hunk 4]
```diff
@@ -723,6 +723,84 @@ const wallet = {
...
+      const pin = `${i}`.padStart(6, '0');
+
+      try {
+        const decrypted = decrypt(data, pin);
+
+        if (decrypted.length > 250) continue;
```
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/wallet.js]: Done in the hathor-wallet-mobile repo

**Review by pedroferreira1 — APPROVED** [review]

**Review by msbrogli — APPROVED** [review]

**Pull request merged**